articles/qos_deadline_liveliness_lifespan.md

```diff
@@ -0,0 +1,188 @@
+
+#### QoS Struct
+
+Minimal changes will need to be made to the QoS struct that is passed into the creation functions for Topics, Services, and Actions.
+A couple new enum values will be added for the Deadline and Liveliness settings and then a couple integers will be added to specify the time values for these policies.
+
+#### Assert Liveliness Functions
+
+Two new functions would need to be added that could be used by the application to explicitly assert liveliness.
+One function to assert liveliness at the Node level and one to assert it at the Topic level.
+While liveliness will also be implicitly assumed just based on sending messages, it will also be able to be explicitly declared by the application.
+These functions will need to be implemented in the rmw layer, the rcl layer, and the language specific client libraries.
+
+#### rcl_wait and rmw_wait
+
+The rcl layer is currently using a WaitSet in order to be informed of events from the rmw layer, such as incoming messages.
+These WaitSets contain lists of several types of conditions, such as timers firing or subscriptions receiving data.
+In order to support new Topic Status Events, a new type will be added to the existing WaitSet and the rmw layer will set them when these events occur.
+
+
+## FAQ
+
+- How does the Deadline policy take into account the additional overhead of ROS (such as deserialization) when determining if a deadline was missed?
+  - As a simplification it is not going to attempt to take into account any ROS overhead. A deadline will be considered missed if the rmw layer does not receive a message by the deadline and not if the user application on top of ROS does not receive it by the deadline.
+
+
+## Open Questions
+
+- How do Deadlines and Liveliness impact Services and Actions?
+  - Actions and Services are slightly different in that they are composed of multiple Topics. It may make sense for a Service to specify a Liveliness policy on its own DataWriter, but it likely doesn't need the same policy on the DataReader.
+  - One option would be to only apply the QoS settings to one or another of the topics. So Liveliness may only apply to the DataWriters in Services and Actions and the Deadlines might apply only to the "update" topic data writer for Actions.
+  - Another option would be to modify the interface to allow for different QoS policies for the different underlying Topics, though this adds more complexity.
+- How do these new QoS settings impact non-DDS based rmw implementations?
+  - While these QoS policies are native to the DDS standard, any non-DDS based rmw layer will not have native support for these policies. What does this mean for those implementations? Will they be forced to implement these features in their rmw_implementation layer to be officially supported?
```

Comment:
It could be optional. If the underlying middelware does not provide it then the rmw implementer has a choice of either providing it in the rmw layer or raising an "unsupported confuguration" exception when a request is made from the rcl layer. I think this would give a good balance of supporting advanced DDS features while allowing other middlewares to be used even if they don't have those features.

Reply:
Yep, I've made them optional by providing recommendations on behavior and allowing for a feedback mechanism from the rmw layer to the application to know that the rmw layer did not apply their full policy if it decides not to throw an exception.